### PR TITLE
Set TCP_USER_TIMEOUT socket option for linux 

### PR DESCRIPTION
**Pull request by yashykt**
Implements https://github.com/grpc/proposal/blob/master/A18-tcp-user-timeout.md

Manually tested

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                           FILE SIZE
 ++++++++++++++ GROWING                                             ++++++++++++++
  +0.0%    +272 [None]                                                 +776  +0.0%
  +9.5%    +352 src/core/lib/iomgr/socket_utils_common_posix.cc        +352  +9.5%
      [NEW]    +348 grpc_set_socket_tcp_user_timeout                       +348  [NEW]
      +2.1%      +4 [Unmapped]                                               +4  +2.1%
  +1.0%     +32 src/core/lib/iomgr/tcp_client_posix.cc                  +32  +1.0%
      +3.1%     +21 grpc_tcp_client_prepare_fd                              +21  +3.1%
       +26%     +11 [Unmapped]                                              +11   +26%
  +0.8%     +16 src/core/lib/iomgr/tcp_server_utils_posix_common.cc     +16  +0.8%
      +2.6%     +23 grpc_tcp_server_prepare_socket                          +23  +2.6%

  +0.0%    +672 TOTAL                                               +1.15Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,953,022      Total (>)      1,952,970

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,007      Total (>)     10,664,807

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by srini100** [comment]
Is this only for Linux? Is there an equivalent in MacOS?

**Comment by yashykt** [comment]
I agree that there should be a way to configure it (including disabling it). I will put out a C++ specific gRFC for that. I am keeping that PR separate from this because that can only go in after the gRFC is approved.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,953,022      Total (>)      1,952,970

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,261      Total (>)     10,665,046

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/iomgr/socket_utils_common_posix.cc [hunk 1]
```diff
@@ -242,7 +242,7 @@ grpc_error* grpc_set_socket_tcp_user_timeout(int fd, int val) {
     return GRPC_ERROR_CREATE_FROM_STATIC_STRING(
         "Failed to set TCP_USER_TIMEOUT");
   }
-#endif /* GPR_LINUX */
+#endif /* GRPC_HAVE_TCP_USER_TIMEOUT */
```
Comment (srini100) [inline src/core/lib/iomgr/socket_utils_common_posix.cc]: Add an info log if this cannot be set.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by srini100** [comment]
As discussed, let's hold off on this PR and do a gRFC to add channel arg and sort out default and min values.

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,953,107      Total (>)      1,952,970

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,259      Total (>)     10,665,045

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,953,107      Total (>)      1,952,970

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,665,255      Total (>)     10,665,053

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/iomgr/socket_utils_common_posix.cc [hunk 2]
```diff
@@ -222,6 +222,32 @@ grpc_error* grpc_set_socket_low_latency(int fd, int low_latency) {
   return GRPC_ERROR_NONE;
 }
 
+#define DEFAULT_TCP_USER_TIMEOUT 20000 /* 20 seconds */
```
Comment (sreecha) [inline src/core/lib/iomgr/socket_utils_common_posix.cc]: Add the `MILLIS` suffix to the name  i.e `DEFAULT_TCP_USER_TIMEOUT_MILLIS`

File: src/core/lib/iomgr/tcp_client_posix.cc [hunk 3]
```diff
@@ -76,6 +76,8 @@ static grpc_error* prepare_socket(const grpc_resolved_address* addr, int fd,
   if (!grpc_is_unix_socket(addr)) {
     err = grpc_set_socket_low_latency(fd, 1);
     if (err != GRPC_ERROR_NONE) goto error;
+    err = grpc_set_socket_tcp_user_timeout(fd, 0 /* set to gRPC default */);
```
Comment (sreecha) [inline src/core/lib/iomgr/tcp_client_posix.cc]: Can you add support to set this as a channel arg as well ?
Fortunately, the required plumbing is already in place - i.e `channel_args` is passed as an argument to this function and just few lines below , we even have code that parses the args  for `GRPC_ARG_SOCKET_MUTATOR` :)

File: src/core/lib/iomgr/socket_utils_common_posix.cc (hunk 2 above)
Comment (yashykt) [reply src/core/lib/iomgr/socket_utils_common_posix.cc]: changed the suffix to _MS following code at other places

File: src/core/lib/iomgr/tcp_client_posix.cc (hunk 3 above)
Comment (yashykt) [reply src/core/lib/iomgr/tcp_client_posix.cc]: Done

**Comment by yashykt** [comment]
It would be hard to create infrastructure that would bring down interfaces to imitate a bad connection, hence manually tested this.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                           FILE SIZE
 ++++++++++++++ GROWING                                             ++++++++++++++
  +0.0%    +288 [None]                                              +1.38Ki  +0.0%
  +1.7%     +64 src/core/lib/iomgr/socket_utils_common_posix.cc         +64  +1.7%
      [NEW]     +42 grpc_set_socket_tcp_user_timeout                        +42  [NEW]
       +10%     +20 [Unmapped]                                              +20   +10%
      [NEW]      +2 config_default_tcp_user_timeout                          +2  [NEW]
  +1.0%     +32 src/core/lib/iomgr/tcp_client_posix.cc                  +32  +1.0%
      +4.0%     +27 grpc_tcp_client_prepare_fd                              +27  +4.0%
       +13%      +5 [Unmapped]                                               +5   +13%
  +1.6%     +32 src/core/lib/iomgr/tcp_server_utils_posix_common.cc     +32  +1.6%
      +3.4%     +31 grpc_tcp_server_prepare_socket                          +31  +3.4%
      +4.2%      +1 [Unmapped]                                               +1  +4.2%

  +0.0%    +416 TOTAL                                               +1.51Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,956,113      Total (>)      1,955,968

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,844,340      Total (>)     10,843,878

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                           FILE SIZE
 ++++++++++++++ GROWING                                             ++++++++++++++
  +0.0%    +320 [None]                                              +5.36Ki  +0.1%
  +1.7%     +64 src/core/lib/iomgr/socket_utils_common_posix.cc         +64  +1.7%
      [NEW]     +42 grpc_set_socket_tcp_user_timeout                        +42  [NEW]
       +10%     +20 [Unmapped]                                              +20   +10%
      [NEW]      +2 config_default_tcp_user_timeout                          +2  [NEW]
  +1.0%     +32 src/core/lib/iomgr/tcp_client_posix.cc                  +32  +1.0%
      +4.0%     +27 grpc_tcp_client_prepare_fd                              +27  +4.0%
       +13%      +5 [Unmapped]                                               +5   +13%
  +1.6%     +32 src/core/lib/iomgr/tcp_server_utils_posix_common.cc     +32  +1.6%
      +3.4%     +31 grpc_tcp_server_prepare_socket                          +31  +3.4%
      +4.2%      +1 [Unmapped]                                               +1  +4.2%

  +0.0%    +448 TOTAL                                               +5.48Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,059      Total (>)      1,960,914

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,611      Total (>)     10,883,158

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yashykt** [comment]
Known issues : #14128, #15489, #16123 
Not sure about the ruby flake but it's possibly the same as #16587

**Comment by yashykt** [comment]
Thanks for reviewing!

**Pull request merged**